File: wear/src/main/java/com/pdyjak/powerampwear/music_browser/BrowserRVAdapter.java
package com.pdyjak.powerampwear.music_browser;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class BrowserRVAdapter<T> extends RecyclerView.Adapter<ItemViewHolder> {

    @NonNull
    private final ViewHolderFactory mFactory;
    @NonNull
    private final List<T> mItems = new ArrayList<>();

    public BrowserRVAdapter(@NonNull ViewHolderFactory factory) {
        mFactory = factory;
    }

    public void setItems(@NonNull List<T> items) {
        if (mItems.equals(items)) return;
        mItems.clear();
        mItems.addAll(items);
        notifyDataSetChanged();
    }

    @NonNull
    public List<T> getItems() {
        return Collections.unmodifiableList(mItems);
    }

    @Override
    public ItemViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return mFactory.createViewHolder(parent);
    }

    @Override
    public void onBindViewHolder(ItemViewHolder holder, int position) {
        holder.bindWith(mItems.get(position));
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }
}

File: wear/src/main/java/com/pdyjak/powerampwear/music_browser/BrowserRVAdapter.kt
package com.pdyjak.powerampwear.music_browser

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup

import java.util.ArrayList
import java.util.Collections

class BrowserRVAdapter<T>(private val mFactory: ViewHolderFactory)
    : RecyclerView.Adapter<ItemViewHolder>() {

    private val mItems = ArrayList<T>()

    var items: List<T>
        get() = Collections.unmodifiableList(mItems)
        set(items) {
            if (mItems == items) return
            mItems.clear()
            mItems.addAll(items)
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
        return mFactory.createViewHolder(parent)
    }

    override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
        val obj = mItems[position]
        holder.bindWith(obj)
    }

    override fun getItemCount(): Int {
        return mItems.size
    }
}
